### ISSUE TITLE
Docker cli lists volume as attached whereas admin cli shows same volume as detached

### ISSUE THREAD

**Issue by ashahi1** [vmware/docker-volume-vsphere#1621]
Test runs on a swarm setup of one master node and two worker nodes.

Test starts a service on swarm worker-1. Then first worker node is shutdown and test expects docker service to start on second worker. After the service successfully starts on the second worker, but one of the verification fails because docker cli shows volume as attached whereas admin cli shows volume as attached.

https://ci.vmware.run/vmware/docker-volume-vsphere/1266

```
Test steps:
Test vDVS usage during failover across different swarm nodes
Test steps:
1. Create a service with volume mounted
2. Verify the service is up and running on one node
3. Verify one container is spawned
4. Verify the volume is in attached status
5. Write data to the volume
6. Shutdown the node on which the service is running
7. Verify the service is restarted on a different node
8. Verify the volume is in attached status    <<<<<<<<<<<<<<<<< Fails here
9. Verify the data from this node
10. Remove the service
11. Verify the service is gone
12. Verify the volume is in detached status
```

Test logs:
```
2017/07/21 15:28:09 Shutting down VM [Ubuntu.1404.vsan]
2017/07/21 15:28:09 Sleep for 5 seconds
2017/07/21 15:28:14 Sleep for 5 seconds
2017/07/21 15:28:19 Sleep for 5 seconds
2017/07/21 15:28:24 Sleep for 5 seconds
2017/07/21 15:28:29 Checking docker service [swarm_test_service_933036] status on VM [192.168.31.241]
2017/07/21 15:28:29 Listing docker service [swarm_test_service_933036] running on VM [192.168.31.241]
2017/07/21 15:28:29 Checking running containers for docker service [swarm_test_service_933036] on docker hosts: [192.168.31.163]
2017/07/21 15:28:29 Containers running on docker host [192.168.31.163]: swarm_test_service_933036.1.i8dtp2ko28psxqz79xv4x1fg5
2017/07/21 15:28:29 Printing Docker volume inspect for volume swarm_test_volume_1046104
2017/07/21 15:28:30 OUT: [
    {
        "Driver": "vsphere:latest",
        "Labels": null,
        "Mountpoint": "/mnt/vmdk/swarm_test_volume_1046104@datastore1",
        "Name": "swarm_test_volume_1046104@datastore1",
        "Options": {},
        "Scope": "global",
        "Status": {
            "access": "read-write",
            "attach-as": "independent_persistent",
            "attached to VM": "Ubuntu.1404.vmfs",
            "attachedVMDevice": {
                "ControllerPciSlotNumber": "160",
                "Unit": "0"
            },
            "capacity": {
                "allocated": "15MB",
                "size": "100MB"
            },
            "clone-from": "None",
            "created": "Fri Jul 21 15:27:54 2017",
            "created by VM": "photon.ga.vmfs",
            "datastore": "datastore1",
            "diskformat": "thin",
            "fstype": "ext4",
            "status": "attached"       <<<<<<<<<<<<<<<<<<<<
        }
    }
]
2017/07/21 15:28:30 Confirming attached status for volume [swarm_test_volume_1046104]
2017/07/21 15:28:30 Fetching full name for volume [swarm_test_volume_1046104] from VM [192.168.31.163]
2017/07/21 15:28:31 Full volume name: [swarm_test_volume_1046104@datastore1]
2017/07/21 15:28:33 Expected Attached VM name is [Ubuntu.1404.vmfs]
2017/07/21 15:28:33 Attached VM name from Docker CLI is [Ubuntu.1404.vmfs]
2017/07/21 15:28:33 Attached VM name from Admin CLI is [detached]         <<<<<<<<<<<<<<<<<
2017/07/21 15:28:33 Destroying volume [swarm_test_volume_1046104]

----------------------------------------------------------------------
FAIL: swarm_test.go:74: SwarmTestSuite.TearDownTest

swarm_test.go:77:
    c.Assert(err, IsNil, Commentf(out))
... value *exec.ExitError = &exec.ExitError{ProcessState:(*os.ProcessState)(0xc4201aa260), Stderr:[]uint8(nil)} ("exit status 1")
... Error response from daemon: unable to remove volume: remove swarm_test_volume_1046104: VolumeDriver.Remove: Failed to remove volume: Error caused by file /vmfs/volumes/575db611-d1ad6cf9-0c74-000c29e3609e/dockvols/11111111-1111-1111-1111-111111111111/swarm_test_volume_1046104.vmdk


----------------------------------------------------------------------
PANIC: swarm_test.go:97: SwarmTestSuite.TestFailoverAcrossSwarmNodes

swarm_test.go:158:
    c.Assert(status, Equals, true, Commentf("Volume %s is not attached", s.volumeName))
... obtained bool = false
... expected bool = true
... Volume swarm_test_volume_1046104 is not attached

... Panic: Fixture has panicked (see related PANIC)
```

Logs:
[vmdk_ops.txt](https://github.com/vmware/docker-volume-vsphere/files/1166906/vmdk_ops.txt)

**Comment by govint** [comment]
From the logs this is what happened,

1. Volume swarm_test_volume_1046104 is created by photon.ga.vmfs
1. The volume is attached by Ubuntu.1404.vsan
1. Then Ubuntu.1404.vsan is killed or shutdown,
1. Ubuntu.1404.vmfs then attaches the volume - it can because Ubuntu.1404.vsan is not powered on and hence disk is not in use. Plus the test to prevent an attach based on KV data is removed now (its only a log now). ESX service disk_attach logs the fact that the disk is attached and proceeds - thats fine.
1. Now the VMListener thread gets the notification that Ubuntu.1404.vsan is powered off.
1. VmListener figures that the volume must be detached, without allowing for the fact that the volume may be attached to a different VM - a.k.a BUG!
1. So, VMListener goes ahead and detaches the disk from the powered-off VM's configuration (which is fine and correct) _and_ sets the KV status to detached (over-writing the fact that the KV has recorded the new attached status of the volume from step (4) above.
1. During this time the test is issuing a _VerifyAttachedStatus_ and two requests show different results, the docker CLI reads the KV just before the VMListener makes the disk as detached, while admin CLI gets the KV after VMListener has updated the KV. Hence admin CLI reported _detached disk_ while docker CLI reported _attached disk_.
1. The test doesn't get the expected result and hence tears down the set up and tries to remove the volume. But the volume is still attached.
1. The old ESX side issue kicks in and when the volume is removed, the disk removal fails because the VM is still using the volume (volume busy) but the KV sidecar file gets removed. Now we have a VM thats got the volume attached and the volume KV file has been removed.

Fix is to have the the VMListener be cognizant of the fact that a volume may be re-attached by the time it gets the powered-off notification and do a _compare and set_ op - the KV is updated _only_ if the KV is still showing the powered-off VM as attaching the volume.

**Comment by govint** [comment]
Fixed via #1626
